### ISSUE TITLE
Native Teleport mobile clients for iOS/Android

### ISSUE THREAD

**Issue by webvictim** [gravitational/teleport#3343]
It’d be awesome if we could write Teleport mobile clients for iOS and Android. The web UI is OK (at least on iOS) for very simple operations, but quickly falls down when you need to scroll or do anything more advanced.

To avoid the hassle of handling terminals, it’s important to bear in mind that there are already lots of SSH clients for both platforms. If we could write an app that lets you log into a cluster, get a certificate and then pass that over to one of the existing clients (assuming at least one of them has certificate support) then that’d be awesome.

It’d also be a nifty marketing tool.

**Comment by shaneshort** [comment]
+1 this would be an excellent addition. Perhaps panic could be convinced to support it in prompt?

**Comment by abeluck** [comment]
Is there any sort of workaround here? We have ipads as remote emergency terminals, and would love to  be able to ssh into machines via teleport with them.

**Comment by webvictim** [comment]
@shaneshort I emailed Panic; they said "Thanks for writing in! I'm afraid support for these certificates isn't currently planned, but I will pass this along to engineering for future consideration."

@abeluck I don't currently know of any workarounds to use SSH certificates via mobile client. The best you could do for web-based SSO logins with Teleport would probably be something like this:
- use a native client which supports port forwarding to connect to an SSH bastion server with a forwarded port (`54321` in this example)
- `eval $(ssh-agent)` on the remote server
- run `TELEPORT_LOGIN_BIND_ADDR=127.0.0.1:54321 tsh login --proxy=....` on the remote server
- open the link from `tsh` on your local device's browser and complete sign-in via SSO
- at that point you should have Teleport certs available on the remote machine, so you can run `tsh ssh` commands and get full Teleport access via the same terminal you're already using.

It isn't that pretty and would introduce a dependency on making sure your SSH bastion is secured with something that isn't SSH certificates, but it would work in a pinch.

**Comment by modem7** [comment]
Even being able to have a web view w/ virtual keyboard support (especially function keys) would be a great step forward.

**Comment by anthosz** [comment]
+1 for this request.
Can be very usefull by example to use the proxy app for tcp applications. (tsh not available :/)

**Comment by dontcrash** [comment]
+1

**Comment by anishchhaparwal** [comment]
+1

**Comment by talnetd** [comment]
+1

**Comment by MarkCupitt** [comment]
Adding a HUGE +1 for this .. I used to use juice on android for ssh access on the golf course, now Im stuck in the office

**Comment by GuideAuBug73** [comment]
+1

**Comment by amirex128** [comment]
+1

**Comment by mta59066** [comment]
+1

**Comment by talnetd** [comment]
+1

**Comment by IconicLight** [comment]
Maybe some form of Teleport Connect for mobile (iOS/Android)?

A dedicated mobile app would be amazing with proper scrolling and keyboard support. At least the servers I'm managing feel cumbersome to do anything on mobile.

**Comment by webvictim** [comment]
## What would you like Teleport to do?

`tsh vnet` would work amazingly well on iOS/Android devices via a "virtual VPN"/VPN-on-demand integration on the phone itself.

## What problem does this solve?

Allows the use of native mobile apps connecting to servers/REST endpoints behind Teleport. This isn't currently possible, and would increase the mass-market appeal of Teleport a lot.

## If a workaround exists, please include it.

Use something like Tailscale, which has mobile clients and VPN integrations.

**Comment by daveneeley** [comment]
> ## If a workaround exists, please include it.
> Use something like Tailscale, which has mobile clients and VPN integrations.

Other not-so-great workarounds - host teleport behind another proxy like nginx. Expose the native apps on the nginx proxy, and host the others behind teleport. 

What about Machine ID with a mobile tbot app? It just phones home in the background to rotate certs, no vnet/vpn. Native apps might have to support custom certs.

**Comment by webvictim** [comment]
Another idea would be an app that runs locally
on the phone and behaves as a HTTPS proxy, renewing and injecting Machine ID certs for authentication.

**Comment by carloscabanero** [comment]
Hi! I'm Carlos from Blink Shell (a terminal app for iOS https://blink.sh). Im looking into this and gotta admit I'm a bit lost on what would be the best way to proceed. I am both interested in authentication, and also what I call "inventories", which would be a filtered list of devices the user may have access to. Can we continue discussing here or can somebody reach out to me on Carlos at blink.sh? Thanks for the help!

**Comment by webvictim** [comment]
@carloscabanero Hey - thanks for the comment.

To work with Teleport, an application would need to implement:
- support for using SSH user certificates for authentication to SSH servers
- support for validating SSH host certificates against the fingerprint of a CA when authenticating to SSH servers

The easiest way to get short-lived SSH certificates to use with Teleport is to use the `tsh` command line tool - unfortunately, we don't (currently) build or provide this for iOS. Having applications that could make use of Teleport certificates to connect over SSH would certainly be useful though.

On the topic of inventory, once logged in, the `tsh ls` command provides a permission-checked view of different SSH server hostnames that the user has access to.

**Comment by carloscabanero** [comment]
Thanks a lot for the guidance @webvictim. Support for user and host certificates is already there, but I think we could make it more straightforward. On getting short-lived certificates and the inventory, is there an API call for this? Is this the expected/recommended way of doing things or does most people just use longer term certificates or other methods? Thanks again

**Comment by jeffellin** [comment]
@carloscabanero I started taking a look at this before I saw this.  creating a tsh command in blink.sh looks doable,  The protocol to make the connection doesn't look too hard.   That being said. I am still having some difficulty getting the the latest release to compile cleanly on my mac.

**Comment by carloscabanero** [comment]
Do you mean the latest release for tsh or for Blink. For Blink I can help you out within our tracker or our Discord, feel free to reach out.

We could try to port the tsh command. There are a few problems with that, not all technical. Our shell is small because we are focused on taking you to the remote asap (so we want to facilitate the connection but not necessarily everything else), keeping the tsh command updated (it is more of a moving target than let's say the `ping` command), porting the command may not be trivial and some shell or system functions may not be there, etc...

My take would be different, more focused: Let's figure out what calls we need to obtain the temp certificates or anything else that we need, and we can embed it as a special "key type" or a special "step" before the connection if needed.

Wdyt?

**Comment by jeffellin** [comment]
@carloscabanero getting the certificates is only half the battle. In fact getting the certs and listing clients that are available to connect was relatively easy compared to managing the ssh connection.  Teleport uses an SSH channel multiplexed over grpc to communicate to the server. I do have a MVP of a tsh terminal that does the full lifecycle written in swift, it gets the cert and then handles the communication with the teleport proxy, I would think MVP would be to see if that command can be integrated into blink using the blink terminal emulation rather than SwiftTerm.   I'd probably suggest some minimal implmeentation of tsh that supports, logging in, obtaining the cert, and setting up an ssh proxy you can use with ProxyCommand

**Comment by carloscabanero** [comment]
Yep, so the grpc is what I was expecting, it is a big deal if you already figured it out, and most of what is left should be just stitching. 

You do not need to worry about the terminal part when connecting to Blink, it works by simulating a terminal device with input and output, and it just expects a main entry point, like any other command. And the shell takes care of instancing, piping, duplicating stuff, all of that. So if you have kinda wrapped your Swift code into a regular CLI command taking input and output, it will be pretty easy to plug it in. That would be step one.

Now, for step 2. Blink offers its own command for SSH, but internally it works with every SSH channel like a service at the code level. This would be interesting because then we would have sftp, the files.app integration and other future components working as well. To do that, we need to integrate the grpc with libssh itself. This shouldn't be hard either, `tsh` is probably piping through Go, or even offering one of those cool internal socket interfaces with reader/writer, etc... In this case I would use regular low level tools, seperating the grpc and the libssh runners, and have them communicate through a descriptor. We could have that through ProxyCommand or a bit more integrated. Maybe.

Let me know how you want to go about this, I'm happy to help any way I can. Ideally we would make the solution open source, but if you are thinking about something else just let me know. And you can also write me to Carlos at blink.sh.

PS: Just thinking out loud. Could we use the same grpc connection to interact with Mosh? That would be cool, but not sure about  what happens when the grpc connection goes down, how mosh can continue UDPing there, etc... Would have to take a deeper look at the solution.

**Comment by vishal-android-freak** [comment]
https://github.com/vishal-android-freak/telemob

I am attempting to build a client. Please try it and give it a shot. It is opensource so feel free to fork or send PRs :)

**Comment by cad0p** [comment]
My fork of https://github.com/cad0p/vvterm has authentication via Face ID working 

See https://x.com/vaipier/status/2083608347467530414?s=46 for demo
